packages/jest-runtime/src/index.ts

```diff
@@ -302,6 +320,83 @@ class Runtime {
     return cliOptions;
   }
 
+  // unstable as it should be replaced by https://github.com/nodejs/modules/issues/393, and we don't want people to use it
+  unstable_shouldLoadAsEsm = Resolver.unstable_shouldLoadAsEsm;
+
+  private async loadEsmModule(
+    modulePath: Config.Path,
+    query = '',
+  ): Promise<VMModule> {
+    const cacheKey = modulePath + query;
+
+    if (!this._esmoduleRegistry.has(cacheKey)) {
+      const context = this._environment.getVmContext!();
+
+      invariant(context);
+
+      if (this._resolver.isCoreModule(modulePath)) {
+        const core = await this._importCoreModule(modulePath, context);
+        this._esmoduleRegistry.set(cacheKey, core);
+        return core;
+      }
+
+      const transformedFile = this.transformFile(modulePath, {
+        isInternalModule: false,
+        supportsDynamicImport: true,
+        supportsStaticESM: true,
+      });
+
+      const module = new SourceTextModule(transformedFile.code, {
+        context,
+        identifier: modulePath,
+        importModuleDynamically: (
+          specifier: string,
+          referencingModule: VMModule,
+        ) =>
+          this.loadEsmModule(
+            this._resolveModule(referencingModule.identifier, specifier),
+          ),
+        initializeImportMeta(meta: JestImportMeta) {
+          meta.url = pathToFileURL(modulePath).href;
+          // @ts-ignore TODO: fill this
+          meta.jest = {};
+        },
+      });
+
+      this._esmoduleRegistry.set(cacheKey, module);
+    }
+
+    const module = this._esmoduleRegistry.get(cacheKey);
+
+    invariant(module);
+
+    return module;
+  }
+
+  async unstable_importModule(
+    from: Config.Path,
+    moduleName?: string,
+  ): Promise<void> {
+    invariant(
+      runtimeSupportsVmModules,
+      'You need to run with a version of node that supports ES Modules in the VM API.',
+    );
+    invariant(
```

Comment:
Can we move this inside `loadEsmModule` to the call to help TS and avoid a 'why is this assumed to be defined' confusion

Reply:
Or I suppose this might be a perf optimization to avoid it being called for every module in the dependency subgraph?

Reply:
didn't think of it as a performance optimization, it's just me writing everything in a single method at first before splitting it up. I don't think perf of `if (thing)` matters much - I assume it'll be called enough to get optimized before it was a problem